File: ds/stack.go
package ds

import "fmt"

type SNode struct {
	Value string
	Next  *SNode
}

type Stack struct {
	Top *SNode
}

func NewStack() *Stack {
	return &Stack{}
}

func (s *Stack) Push(val string) {
	s.Top = &SNode{Value: val, Next: s.Top}
}

func (s *Stack) Pop() string {
	if s.Top == nil {
		return "[STACK_EMPTY]"
	}
	val := s.Top.Value
	s.Top = s.Top.Next
	return val
}

func (s *Stack) Peek() string {
	if s.Top == nil {
		return "[STACK_EMPTY]"
	}
	return s.Top.Value
}

func (s *Stack) IsEmpty() bool {
	return s.Top == nil
}

func (s *Stack) ReadStack() {
	if s.Top == nil {
		fmt.Println("Стек пуст")
		return
	}
	fmt.Println("Содержимое стека (сверху вниз): ")
	for cur := s.Top; cur != nil; cur = cur.Next {
		fmt.Printf("  %s\n", cur.Value)
	}
}

File: stack.cpp
#include "stack.h"
#include <iostream>
#include <fstream>

using namespace std;

Stack::Stack() : top(nullptr) {}

Stack::~Stack() {
    while (top) {
        SNode* temp = top;
        top = top->next;
        delete temp;
    }
}

void Stack::push(string val) {
    SNode* newNode = new SNode{val, top};
    top = newNode;
}

string Stack::peek() const {
    if (!top) return "[STACK_EMPTY]";
    return top->value;
}

string Stack::pop() {
    if (!top) return "[STACK_EMPTY]";
    SNode* temp = top;
    string val = top->value;
    top = top->next;
    delete temp;
    return val;
}

void Stack::readStack() const {
    if (!top) {
        cout << "Стек пуст." << endl;
        return;
    }
    cout << "Стек (верх -> низ): ";
    SNode* curr = top;
    while (curr) {
        cout << curr->value;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << endl;
}

bool Stack::isEmpty() const {
    return top == nullptr;
}
